flint-and-steel: trigger node on_ignite, add it to tnt, gunpowder, coal

Ferk opened the pull request:
This changes the flint-and-steel logic to trigger an `on_ignite` callback from the nodedef, and moves the previous logic to the `on_ignite` of each of the nodes that flint-and-steel acted on: tnt, gunpower and coalblock.

This way flint-and-steel is more flexible and mods will be able to add their own nodes that can be ignited by flint-and-steel without re-implementing the `on_use` of the tool.

~~It also allows the tnt and gunpowder to be set flammable since fire would now ignite them.~~ (see #1348)

One slight change of behavior is that now when creating a permanent flame ~~both the coalblock and the place where the permanent flames will be placed~~ only the coalblock is checked for protection, instead of the place where the flame is placed.

paramat commented:
> when creating a permanent flame both the coalblock and the place where the permanent flames will be placed are checked for protection

This is fine.

sofar commented:
I'd keep the protection check out of any `on_burn()` callback: natural fire should propagate without permission checks, it's user actions that should be checked.

sofar commented:
:+1: this was overdue (I think I even proposed to write this a while back), and both a nice simplification as well as making things much more flexible.

Ferk commented:
Would it be a good idea to trigger an `on_burn` for entities hit by flint-and-steel too?
Then you could burn up itemstacks.

sofar commented:
Handling entities would be nice, could do so in a separate PR, though, since you'll likely want it to be an `object:on_burn()` type callback.

Ferk commented:
> I'd keep the protection check out of any `on_burn()` callback: natural fire should propagate without permission checks, it's user actions that should be checked.

Do you mean that I should remove the check above the coalblock before placing the permanent flame?
Well.. since the permanent flame is not an igniter and we check for "air" I suppose it might not be much of a grief if someone wanted to abuse it.

I'm not sure about making the coalblock flammable by natural fire, though. I think this could cause problems because unlike tnt or gunpowder, the coalblock does not replace itself when burn. It continues there and it could potentially get its `on_burn` triggered again and again if placed next to lava.
So, unless someone has some good idea it's probably best to leave it non-flammable.

Ferk commented:
@sofar, the coalblock does have a permission check inside its callback (in addition to the one in the flint-and-steel). I was wondering if the check on the flame position is really needed or if maybe the one in the coalblock position is enough.

sofar commented:
Ah, IC, I'd remove the one inside the callback then.

paramat commented:
Permanent flame is an igniter.

paramat commented on mods/fire/init.lua:
```diff
@@ -95,38 +95,18 @@ minetest.register_tool("fire:flint_and_steel", {
+
+			local player_name = user:get_player_name()
+			if nodedef.on_burn
+				and not minetest.is_protected(pt.under, player_name) then
```
Extra indent needed to distinguish the line from the line below.

paramat commented on mods/tnt/init.lua:
```diff
@@ -538,7 +541,7 @@ function tnt.register_tnt(def)
 			tiles = {tnt_top, tnt_bottom, tnt_side},
 			is_ground_content = false,
-			groups = {dig_immediate = 2, mesecon = 2, tnt = 1},
+			groups = {dig_immediate = 2, mesecon = 2, tnt = 1, flammable = 5},
```
SInce it's already burning best not make it flammable, the flame will hover above and not look good anyway.

Ferk replied:
This is not the {{tnt:tnt_burning}} variant, this is the normal {{tnt:tnt}}, the burning variant is not flammable. The point of this is to make the normal tnt able to get started when there's an igniter next to it (thanks to {{on_burn}}).

But since I'm gonna change it to on_ignite, I'll leave that for another PR, as with the gunpowder.

paramat commented on mods/tnt/init.lua:
```diff
@@ -404,7 +404,7 @@ minetest.register_node("tnt:gunpowder", {
 		fixed = {-1/2, -1/2, -1/2, 1/2, -1/2+1/16, 1/2},
 	},
-	groups = {dig_immediate = 2, attached_node = 1, connect_to_raillike = minetest.raillike_group("gunpowder")},
+	groups = {dig_immediate = 2, attached_node = 1, flammable = 5, connect_to_raillike = minetest.raillike_group("gunpowder")},
```
Not sure about flammable, it would be weird for it to burn without changing neighbours to 'gunpowder burning'. Perhaps leave it non-flammable while we consider interaction of fire and tnt stuff.

Ferk replied:
I'm not sure what do you mean.
It does change the neighbour gunpowder to 'gunpowder burning', that was the point of making it flammable, the {{on_burn}} gets triggered and the gunpowder turns into gunpowder burning which then turns its neighbors.

But I think I'm gonna remove this and leave it for another PR, since I'm gonna replace the {{on_burn}} with {{on_ignite}} on this one.

paramat commented:
> Ah, IC, I'd remove the one inside the callback then.

I agree.

paramat commented:
Not sure about using 'on burn' here, that is a callback for flammable nodes that have finished burning and would normally be removed.
Perhaps this PR should use a new function 'on_lighted'?

Ferk commented:
You are right, not sure why i thought otherwise.
That's even more reason against making the coal block flammable. And
perhaps leave the additional permission check for the flame.

paramat commented:
The original 'on burn' callback could have been more clearly named, it is more like 'on burnt'.

Ferk commented:
My previous commentary was answering to your "Permanent flame is an igniter.".... I'm not sure why it took 1 whole day for github to post it, maybe because I used email for it.

But still, you've got a point about the `on_burn`, I guess if mods are already using that one in ways that wouldn't be expected for a flint_and_steel, it wouldn't hurt to make it a different callback.

But `on_light` might be confusing as it might look like it's somehow related to lightning or illumination. I think I'll use `on_ignite`, unless someone comes up with a better name.

I'll still like to have the tnt and gunpowder get activated by fire and lava, though. But that's starting to look like it would be more appropriate for another PR.

Ferk commented:
* changed the callback name to on_ignite (name still open for discussion)
* added documentation for new function
* removed flammable group fron tnt and gunpowder (without on_burn it doesn't make sense)
* removed second protection check for coalblock inside the callback
* fixed indentation style

paramat commented:
Not sure about 'on ignite', this can be confused with ignition by fire, how about one extra word to make it clear, 'on player ignite' or something?

Ferk commented:
It would be interesting if in the future there were other ways to trigger it that not necessarily need a player (I'm thinking gunpowder, for example, so it can be used to activate other things instead of having tnt hardcoded, you could use a circuit of gunpowder to light up all your coalblocks, as if it was a simplified version of mesecons). So I'd rather not imply "player" in the name.

But I'm fine with changing the name if ignite is misleading. It sounds ok to me but I'm not a native english speaker and I wouldn't be able to tell if ignite has different connotation. Maybe "on_spark"? "on_firestart"?

paramat commented:
Perhaps 'on ignite' is okay, since it is different from 'on burn' which is a call when the node has finished burning up. So leave it as it is for now and we'll see what others thiink.

paramat commented on game_api.txt:
```diff
@@ -227,6 +227,12 @@ New node def property:
 
+ `on_ignite(pos, igniter)`
+
+  * Called when a Flint and steel tool is used in a node.
```
'On' instead of 'in':
`* Called when a Flint and steel tool is used on a node.`

paramat commented on mods/fire/init.lua:
```diff
@@ -95,38 +95,18 @@ minetest.register_tool("fire:flint_and_steel", {
+					and not minetest.is_protected(pt.under, player_name) then
+				nodedef.on_ignite(pt.under, user)
+			elseif minetest.get_item_group(node_under, "flammable") >= 1
+					and not minetest.is_protected(pt.above, player_name) then
```
Ah nevermind, i see this is an optimised way to avoid fire if either under or above are protected.

paramat replied:
Yes both need checking, otherwise yo ucould ignite a protected node if the flame pos is outside protection.

paramat replied:
In 'on ignite', 'under' is checked here, and the flame pos is checked in the node's own 'on ignite' function.

Ferk replied:
if 'node.on_ignite' is not defined, 'under' won't be checked, only 'above' will (and that only if the node is flammable).

The right-side expression of an `and` operator doesn't get evaluated if the left-side expression is already false, `and` returns false already as soon as the first expression is false.

paramat replied:
Sorry, that was unclear, i meant:
For the case of a node with 'on ignite', 'under' is checked on line 105 and flame pos is checked in the node's 'on ignite' function, so both positions are checked.
So i'm suggesting that for nodes without 'on ignite' we should do the same.

Ferk replied:
Ah, I think you meant that in the case a flammable node has an on_ignite, if both positions are protectewd both will be checked.

Yes, this was like needed because they were different positions. If they are the same position then it should be factored out, of course.

paramat commented:
Edited my line comments, please reread.

paramat commented on mods/fire/init.lua:
```diff
@@ -95,38 +95,19 @@ minetest.register_tool("fire:flint_and_steel", {
+			end
+			if nodedef.on_ignite then
+				nodedef.on_ignite(pt.under, user)
+			elseif minetest.get_item_group(node_under, "flammable") >= 1 then
```
This new code is better. Still thinking we should check 'pt.above' here.

paramat replied:
> Also, it makes little difference since fire spread in the proximity would reach the protected nodes anyway if fire is enabled and the protected area is flammable.

Oh i see, this is ok then.

Ferk replied:
Oh sorry.. you meant that we should check pt.above for protection? isn't that what the code did before? I thought you were asking me to remove that and only check pt.under, which I think makes sense and simplifies things.

paramat commented:
Added reply to outdated line cmments.

paramat commented:
Ok this seems fine.
This leads to the question: is there any point checking the flame pos in the coalblock's 'on ignite' function?

paramat commented:
Currently i'm thinking 'no' to that question.

Ferk commented:
> is there any point checking the flame pos in the coalblock's 'on ignite' function?

No.. actually I though I had removed that check (both you and sofar requested for it before, actually). It should be no longer there now.

paramat commented:
Well deserved :+1:

rubenwardy commented on game_api.txt:
```diff
@@ -227,6 +227,12 @@ New node def property:
 
+ `on_ignite(pos, igniter)`
+
+  * Called when a Flint and steel tool is used on a node.
```
Mods may define their own igniting tools, so "flint and steel (or a mod defined ignitor)"

rubenwardy replied:
You should also mention that this stops a flame from being spawned - the mod would need to place their own flame

Ferk replied:
Updated it with your suggestions, is it ok?

sofar commented:
:+1:

paramat commented:
c824d6918226cfb88bd1d843b7108d980813ccca
